src/engine/channels/enginedeck.cpp

```diff
@@ -118,31 +120,66 @@ EngineDeck::~EngineDeck() {
 }
 
 #ifdef __STEM__
+void EngineDeck::addStemHandle(const ChannelHandleAndGroup& stemHandleGroup) {
+    m_stems.emplace_back(ChannelHandleAndGroup(stemHandleGroup.handle(), stemHandleGroup.name()));
+    m_stemsGainCache.push_back(CSAMPLE_GAIN_ONE);
+    if (m_pEffectsManager != nullptr) {
+        m_pEffectsManager->registerInputChannel(stemHandleGroup);
+    }
+}
+
 void EngineDeck::processStem(CSAMPLE* pOut, const int iBufferSize) {
-    int stemCount = m_pBuffer->getChannelCount() / mixxx::kEngineChannelOutputCount;
+    mixxx::audio::ChannelCount chCount = m_pBuffer->getChannelCount();
+    VERIFY_OR_DEBUG_ASSERT(m_stems.size() <= chCount &&
+            m_stemMute.size() <= chCount && m_stemGain.size() <= chCount) {
+        return;
+    };
+    mixxx::audio::SampleRate sampleRate = mixxx::audio::SampleRate::fromDouble(m_sampleRate.get());
+    int stemCount = chCount / mixxx::kEngineChannelOutputCount;
+    int numFrames = iBufferSize / mixxx::kEngineChannelOutputCount;
     auto allChannelBufferSize = iBufferSize * stemCount;
     if (m_stemBuffer.size() < allChannelBufferSize) {
         m_stemBuffer = mixxx::SampleBuffer(allChannelBufferSize);
     }
     m_pBuffer->process(m_stemBuffer.data(), allChannelBufferSize);
 
-    // TODO(XXX): process effects per stems
-    SampleUtil::clear(pOut, iBufferSize);
-    const CSAMPLE* pIn = m_stemBuffer.data();
-    for (int i = 0; i < iBufferSize; i += mixxx::kEngineChannelOutputCount) {
+    CSAMPLE* pIn = m_stemBuffer.data();
+    EngineEffectsManager* pEngineEffectsManager = m_pEffectsManager->getEngineEffectsManager();
+    if (pEngineEffectsManager != nullptr) {
+        GroupFeatureState featureState;
+        collectFeatures(&featureState);
         for (int stemIdx = 0; stemIdx < stemCount;
                 stemIdx++) {
-            if (m_stemMute[stemIdx]->toBool()) {
-                continue;
-            }
-            float gain = static_cast<float>(m_stemGain[stemIdx]->get());
-            pOut[i] += pIn[stemCount * i + mixxx::kEngineChannelOutputCount * stemIdx] * gain;
-            pOut[i + 1] +=
-                    pIn[stemCount * i +
-                            mixxx::kEngineChannelOutputCount * stemIdx + 1] *
-                    gain;
+            int chOffset = stemIdx * mixxx::audio::ChannelCount::stereo();
```

Comment:
This is now the final version of `EngineDeck::processStem`!

Reply:
I am struggling to undertand the code here with all the diffrent Layouts and in and out buffers. 
An output to one function is an input to another. 
Can you go through it and try to rename the variables more verbose?
You may also codider alias pointers that have the correct name for the bufer with the current content. 
Also // LRL .. viualization would be helpfull.

Reply:
Let me know if it easier to understand now! Feel free to suggest better phrasing for the comments.

Reply:
Perfect. Thank you.